### PR TITLE
FlippenZap

### PR DESCRIPTION
**Pull request by rayeaster**
for https://github.com/Badger-Finance/ebtc/issues/709

Assumption: There is a `stETH<->eBTC` pair in Balancer V2 as on-chain liquidity and `FlippenZap` user will grant it approval as `PositionManager`

- support `enterLongEth` by creating leveraged CDP (leverage in a safe range of `[2X, 5X]`) or plain (no leverage) CDP Zap-in with given initial principal of `stETH`
- support `enterLongBtc` by simply swap initial principal of `stETH` to eBTC using above liquidity
- add a `MockBalancer` for test purpose
- add test for a single `enterLongEth` Zap happy path
- add test for triple `enterLongEth` Zaps
- add test for multiple users for  `enterLongEth` Zaps
- deployed & verified on [sepolia testnet](https://sepolia.etherscan.io/address/0xe4524de47ff7e5d862cd509ce12e4547e51d3302#readContract)

### CODE REVIEW

File: packages/contracts/contracts/FlippenZap.sol [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+            (address, uint256, uint256)
+        );
+        uint256 _expectedDebt = amount + fee;
+
+        // swap eBTC to collateral
+        uint256 _outAmount = _singleSwapViaBalancer(
+            address(ebtcToken),
+            address(ebtcCollToken),
+            amount,
+            ((_collToLoan * (MAX_SLIPPAGE - slippage)) / MAX_SLIPPAGE)
```
Comment (aviggiano) [inline packages/contracts/contracts/FlippenZap.sol:205]: I think slippage should be a parameter of `enterLongEth`

File: packages/contracts/contracts/FlippenZap.sol [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        // mints eBTC against stETH collateral and
+        // sells the eBTC back for stETH.
+
+        bytes32 _newCdp;
+        uint256 _collToLoan = _leverage > 1 ? (_leverage - 1) * _initialStETH : 0;
+        uint256 _expectedColl = _collToLoan + _initialStETH;
+        uint256 _expectedDebt;
+
+        // transfer initial collateral
+        ebtcCollToken.transferFrom(msg.sender, address(this), _initialStETH);
```
Comment (aviggiano) [inline packages/contracts/contracts/FlippenZap.sol:112]: `ebtcCollToken` is stETH, right?

So this needs to account for the ["1-wei corner case"](https://docs.lido.fi/guides/lido-tokens-integration-guide/#1-wei-corner-case), i.e., the `transferFrom` needs to have a before/after balance to do the accounting properly.

File: packages/contracts/contracts/FlippenZap.sol [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        require(_initialStETH > MIN_NET_STETH_BALANCE, "FlippenZap: initial principal too small");
+        require(_leverage <= MAX_REASONABLE_LEVERAGE, "FlippenZap: _leverage is too big");
+        require(_leverage > 0, "FlippenZap: _leverage is too small");
+
+        // under the covers:
+        // mints eBTC against stETH collateral and
+        // sells the eBTC back for stETH.
+
+        bytes32 _newCdp;
+        uint256 _collToLoan = _leverage > 1 ? (_leverage - 1) * _initialStETH : 0;
```
Comment (aviggiano) [inline packages/contracts/contracts/FlippenZap.sol:107]: I am not sure this calculation is correct. 

Maybe it is.

I always get confused about the role of the gas stipend in some calculations, so I'm just thinking out loud:

Assume the user has 10 stETH and the gas stipend is 1 stETH, for the sake of simplicity. Then the user wants to leverage 5x. This means they effectively have 9 stETH at their disposal. Shouldn't this be the number multiplied by 5x, i.e., 45 stETH as a result? Not 50 stETH?

**Review by aviggiano — COMMENTED** [review]
Some random comments

File: packages/contracts/contracts/FlippenZap.sol (hunk 2 above)
Comment (rayeaster) [reply packages/contracts/contracts/FlippenZap.sol:112]: Good catch ser. Fixed as suggested

File: packages/contracts/contracts/FlippenZap.sol (hunk 3 above)
Comment (rayeaster) [reply packages/contracts/contracts/FlippenZap.sol:107]: Yep, the value of liquidator stipend should be reduced from initial principal before the calculation for flashloan. Fixed.

File: packages/contracts/contracts/FlippenZap.sol (hunk 1 above)
Comment (rayeaster) [reply packages/contracts/contracts/FlippenZap.sol:205]: We may want to hide the implementation details (like the flashloan involved) from the end user so UI could focus clearly on convenient one-click Zap-in for `enterLongEth`

@dapp-whisperer please share your insight on this.

**Pull request closed without merging**